contents/regex/Regex.md

```diff
@@ -0,0 +1,316 @@
+
+The problem occurs because `.` matches any character. This also includes the comma separating the group. This is what happens when the first set of `(.+,)` is parsed:
+
+```
+[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+^
+[.........................,
+[.+                       ,
+```
+
+The first group keeps everything as its data group. Now, it tries to find the second `(.+,)` in ` 10]`, but fails, so it backtracks:
+
+```
+[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+^
+[......................,..,
+[.+                    ,.+,
+```
+
+This time the first and second groups are matched. Now, it tries to find the **third** `(.+,)` in ` 10]`, but fails, so it backtracks yet again. So we move the first `(.+,)` back:
+
+```
+[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+^
+[...................,
+[.+                 ,
+```
+
+Now, what happens? Well, the second `(.+,)` is greedy, so it actually matches the last comma instead!
+
+```
+[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+^
+[...................,.....,
+[.+                 ,.+   ,
+```
+
+As we can see, by chaining repeated groups, it actually causes cascading levels of backtracking, which leads to very poor performance, even with short strings. For this case, we can observe that each comma is either included in the regex or not included in the match, and every possible combination of included and not included combinations will appear at least once. That means the complexity is equal to the number of possible ways n commas can be either included or not included, or O(2<sup>n</sup>). This is a Very Bad Thing™!
+
+This problem can be exacerbated by the one mentioned previously. Imagine if your text looked like this:
+
+```
+[1, 2, 3, 4, 5, 6, 7, 8, 9, 10] some unrelated words, blah blah (200 more redacted)
+[...................................................,
+```
+
+The extra parts of the string will be matched each time, and we can only imagine how long that will take.
+
+So, how do we prevent this?
+
+* Explicitly exclude the delimter.
+
+> Match 10 comma separated values in an array: `\[([^,]+,){9}([^,]+)\]` instead of `\[(.+,){9}(.+)\]`<br>
+
+Here, instead of matching all characters (`.`), we match all characters except a certain character (`[^,]`). This allows the regex to capture in one parse successfully. Some other common examples are HTML/XML tags (`<[^>]*>` instead of `<.+>`) and bracketed expressions (`\([^\)]+\)` instead of `\(.+\)`).
+
+* Always narrow the scope as much as possible.
+
+> Matching a name: `[A-z]+ [A-z]+` instead of `.+ .+`<br>
+
+A great tip for improving Regex performance is to specify your match as clearly as possible. This reduces the potential of unwanted matchings and unnecessary backtracking.
+
+* Avoid greedy operators if possible.
+
+> Matching a date: `\d{1,2}-\d{1,2}-\d{2,4}` instead of `.*-.*-.*`<br>
+
+If you know your data will only be a few characters long, try your best to avoid using `*` or `+`. Limiting to a number of characters would reduce the amount of backtracking if any.
```

Comment:
You may want to make mentions to lazy quantifiers here. It's often the solution to the catastrophic backtracking.